Sources/web3swift/Contract/EthereumContract.swift

```diff
@@ -6,137 +6,140 @@
+            var array = methods[key] ?? []
+            array.append(value)
+            methods[key] = array
+        }
 
-    public struct EventFilter {
-        public var parameterName: String
-        public var parameterValues: [AnyObject]
-    }
+        for case let .function(function) in abi where function.name != nil {
+            appendFunction(function.name!, function)
+            appendFunction(function.signature, function)
+            appendFunction(function.methodString.addHexPrefix().lowercased(), function)
 
-    public var methods: [String: ABI.Element] {
-        var toReturn = [String: ABI.Element]()
-        for m in self._abi {
-            switch m {
-            case .function(let function):
-                guard let name = function.name else {continue}
-                toReturn[name] = m
-            default:
-                continue
+            /// ABI cannot have two functions with exactly the same name and input arguments
+            if (methods[function.signature]?.count ?? 0) > 1 {
+                fatalError("Given ABI is invalid: contains two functions with possibly different return values but exactly the same name and input parameters!")
             }
         }
-        return toReturn
-    }
-
-    public var constructor: ABI.Element? {
-        var toReturn: ABI.Element? = nil
-        for m in self._abi {
-            if toReturn != nil {
-                break
-            }
-            switch m {
-            case .constructor(_):
-                toReturn = m
-                break
+        return methods
+    }()
+
+    private(set) public lazy var allMethods: [ABI.Element.Function] = {
+        return methods.filter { pair in
+            let data = Data.fromHex(pair.key)
+            return data?.count == 4
+        }.values.flatMap { $0 }
+    }()
+
+    private(set) public lazy var events: [String: ABI.Element.Event] = {
+        var events = [String: ABI.Element.Event]()
+        for case let .event(event) in abi {
+            events[event.name] = event
+        }
+        return events
+    }()
+
+    private(set) public lazy var allEvents: [ABI.Element.Event] = {
+        return Array(events.values)
+    }()
+
+    private(set) public lazy var constructor: ABI.Element.Constructor = {
+        for element in abi {
+            switch element {
+            case let .constructor(constructor):
+                return constructor
             default:
                 continue
             }
         }
-        if toReturn == nil {
-            let defaultConstructor = ABI.Element.constructor(ABI.Element.Constructor(inputs: [], constant: false, payable: false))
-            return defaultConstructor
-        }
-        return toReturn
+        return ABI.Element.Constructor(inputs: [], constant: false, payable: false)
+    }()
+
+    public init(abi: [ABI.Element]) {
+        self.abi = abi
     }
 
-    public var events: [String: ABI.Element.Event] {
```

Comment:
Would I'll be wrong if I say that everything above this line is just a code refactoring?

Reply:
I'd say there were refactoring and bug fixes.

Reply:
To be clear everything related to `var methods: [String: [ABI.Element.Function]]` was primarily a bug fix.
The rest is refactoring.